File: Leetcode/剑指offer/easy/20210914/maxSubArray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int dp[nums.size()];
        int res=INT_MIN;
        for(int i=0;i<nums.size();i++){
            if(i==0){
                dp[i]=nums[i];
            }else{
                dp[i]=max(dp[i-1]+nums[i],nums[i]);
            }
            res=max(res,dp[i]);
        }
        return res;
    }
};

File: Leetcode/程序员面试金典/20211014/maxSubArray.go
package _0211014

func maxSubArray(nums []int) int {
	dp := make([]int, len(nums))
	res := 0
	for i := 0; i < len(nums); i++ {
		if i == 0 {
			dp[i] = nums[i]
			res = nums[i]
			continue
		}
		dp[i] = max(dp[i-1]+nums[i], nums[i])
		res = max(res, dp[i])
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
